### PR TITLE
#144 integrate with HiveMQ client feature complete

### PR DESCRIPTION
**Pull request by skobow**
Integration of HiveMQ MQTT client feature complete with running tests. Configuration will be changed and adopted to HiveMQ client. Documentation missing.

### CODE REVIEW

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (timyates) [inline mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java:1]: Do you have Graal configuration for the hivemq client?

I copied `test-suite-mqttv3-graal` to `test-suite-hivemq-mqttv3-graal`, added

`
@Property(name = "mqtt.client.mqtt-version", value = "3")
`

To the `MqttTest` 

and changed `test-suite-hivemq-mqttv3-graal/build.gradle.kts` so its dependencies were

```
dependencies {
    testImplementation(projects.micronautMqttHivemq)
}
```

Running `./gradlew :test-suite-hivemq-mqttv3-graal:test` worked 😎 👍 

But switching to use Graal 17 and running `./gradlew :test-suite-hivemq-mqttv3-graal:nativeTest` failed with a Graal reflection error:

https://ge.micronaut.io/s/yf7uam4fmseg6/console-log?task=:test-suite-hivemq-mqttv3-graal:nativeTest

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                topicMap.entrySet().stream().map(topicEntry -> Mqtt3Subscription.builder()
+                    .topicFilter(topicEntry.getKey())
+                    .qos(Objects.requireNonNull(MqttQos.fromCode(topicEntry.getValue())))
+                    .build())
+            ).build();
+
+        client.subscribe(mqttSubscribe, mqtt3Publish -> {
+                LOG.trace("Received message: {}", new String(mqtt3Publish.getPayloadAsBytes()));
+
+                // TODO: populate MqttMessage properly
```
Comment (timyates) [inline mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java:79]: Does this need addressing?

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v5/client/Mqtt5ClientAdapter.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                topicMap.entrySet().stream().map(topicEntry -> Mqtt5Subscription.builder()
+                    .topicFilter(topicEntry.getKey())
+                    .qos(Objects.requireNonNull(MqttQos.fromCode(topicEntry.getValue())))
+                    .build())
+            ).build();
+
+        client.subscribe(mqttSubscribe, mqtt5Publish -> {
+                LOG.trace("Received message: {}", new String(mqtt5Publish.getPayloadAsBytes()));
+
+                // TODO: populate MqttMessage properly
```
Comment (timyates) [inline mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v5/client/Mqtt5ClientAdapter.java:81]: Does this need addressing?

**Review by timyates — CHANGES_REQUESTED** [review]
Really really nice! 😎 👍 

Couple of requests

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java (hunk 1 above)
Comment (timyates) [reply mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java:1]: Oooooh, so I found this PR from last week:

https://github.com/hivemq/hivemq-mqtt-client/pull/578

If I add

```diff
diff --git a/mqtt-hivemq/src/main/resources/META-INF/native-image/io.micronaut.mqtt/mqtt-hivemq/reflect-config.json b/mqtt-hivemq/src/main/resources/META-INF/native-image/io.micronaut.mqtt/mqtt-hivemq/reflect-config.json
index e69de29..c82657e 100644
--- a/mqtt-hivemq/src/main/resources/META-INF/native-image/io.micronaut.mqtt/mqtt-hivemq/reflect-config.json
+++ b/mqtt-hivemq/src/main/resources/META-INF/native-image/io.micronaut.mqtt/mqtt-hivemq/reflect-config.json
@@ -0,0 +1,18 @@
+[
+  {
+    "name": "org.jctools.queues.BaseSpscLinkedArrayQueueConsumerField",
+    "fields": [
+      {
+        "name": "consumerIndex"
+      }
+    ]
+  },
+  {
+    "name": "org.jctools.queues.BaseSpscLinkedArrayQueueProducerFields",
+    "fields": [
+      {
+        "name": "producerIndex"
+      }
+    ]
+  }
+]
```

Then the Graal native tests pass

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+/**
+ * Adapter class for MQTT 3 HiveMQ async client.
+ *
+ * @author Sven Kobow
+ * @since 3.0.0
+ * @see MqttClientAdapter
+ * @see Mqtt3AsyncClient
+ */
+@Singleton
+@Requires(property = "mqtt.client.mqtt-version", value = "3")
```
[the remark this answers is not in the document]
Comment (skobow) [reply mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v3/client/Mqtt3ClientAdapter.java:50]: Done. I additionally set MQTT v5 to be the default even if the property is not provided.

File: mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v5/client/Mqtt5ClientAdapter.java (hunk 3 above)
Comment (skobow) [reply mqtt-hivemq/src/main/java/io/micronaut/mqtt/hivemq/v5/client/Mqtt5ClientAdapter.java]: Added population of missing properties. Done.

**Comment by sdelamo** [comment]
@skobow are ok you of us merging this PR into the base branch and review adjusting the other PR?

**Comment by skobow** [comment]
> 

@sdelamo you mean PR #256? I am absolutely fine with merging?

**Comment by skobow** [comment]
> 

@timyates anything I can assist with? My goal was that v3 and v5 could possibly use the same configuration. Of course there are some differences but non should be conflicting. But maybe I just didn't get the issue right.

**Comment by timyates** [comment]
@skobow No, we should be fine...  Our annotation processor that generates docs didn't expect multiple configurations in the same module with the same prefix...  We can take it from here :-)

**Pull request merged**